### PR TITLE
Fallback for keepalive js

### PR DESCRIPTION
**Pull request by andrepereiradasilva**
Pull Request for Issue https://github.com/joomla/joomla-cms/issues/16375 (only keepalive fallback part).

### Summary of Changes

The keepalive js fallback is not working correctly when there is no `system.keepalive` json. This PR intends to add better fallback for those wild scenarios.

### Testing Instructions

Pre-requisites:
- Use latest joomla staging with a login module in the frontend (so keepalive is called)
- In global config set joomla session to 1 minute (for faster test) and System Cache to "ON - Progressive caching"

Reproducing the issue:
- Clear joomla cache
- Go to frontend - not logged, reload the page two times (Ctrl+F5)
- Confirm there is a **js error** in the Chrome Debug Console "Console" tab (F12)

Test if proposed keepalive fallback works:
- Apply patch https://github.com/joomla/joomla-cms/pull/16488 (if not merged) and them this patch
- Repeat steps in "Reproducing the issue", but now confirm there is **no js error** in the Chrome Debug Console "Console" tab (F12)
- Wait one minute and also confirm keepalive in called correctly in Chrome Debug Console "Network" tab

### Expected result

Keepalive should always work even if there is no `system.keepalive` json.

### Actual result

Keepalive does not work in this scenario because of a js error.

### Documentation Changes Required

@dgt41 @Fedik can you also check this one and confirm there is no problem in adding the joomla install uri paths trough script options in behavior.core, ie, all the pages that use the core bahaviour (this would also be used in https://github.com/joomla/joomla-cms/pull/15529).

### CODE REVIEW

**Comment by Fedik** [comment]
@andrepereiradasilva I did not tested, but looks good to me

File: media/system/js/keepalive-uncompressed.js [hunk 1]
```diff
@@ -17,12 +17,20 @@
 	document.addEventListener('DOMContentLoaded', function() {
 
 		var keepaliveOptions  = Joomla.getOptions('system.keepalive'),
-		    keepaliveUri      = keepaliveOptions.uri ? keepaliveOptions.uri.replace(/&amp;/g, '&') : window.location.pathname,
-		    keepaliveInterval = keepaliveOptions.interval ? keepaliveOptions.interval : 45 * 1000;
+		    keepaliveUri      = keepaliveOptions && keepaliveOptions.uri ? keepaliveOptions.uri.replace(/&amp;/g, '&') : '',
+		    keepaliveInterval = keepaliveOptions && keepaliveOptions.interval ? keepaliveOptions.interval : 45 * 1000;
+
+		// Fallback in case no keepalive uri was found.
+		if (keepaliveUri === '')
```
Comment (Fedik) [inline media/system/js/keepalive-uncompressed.js]: this check only for empty string, for check all "false" cases use `if (!keepaliveUri)`

Comment (andrepereiradasilva) [inline media/system/js/keepalive-uncompressed.js]: yes but i want to strict check for empty string :) see above
```js
keepaliveUri      = keepaliveOptions && keepaliveOptions.uri ? keepaliveOptions.uri.replace(/&amp;/g, '&') : '',
```

Comment (Fedik) [inline media/system/js/keepalive-uncompressed.js]: okay 😉

**Comment by dgt41** [comment]
@Fedik can you also update the spinning logo to move this code instead of the data-attr (https://github.com/joomla/joomla-cms/pull/15529)?

File: libraries/cms/html/behavior.php [hunk 2]
```diff
@@ -88,6 +88,10 @@ public static function core()
 		}
 
 		JHtml::_('script', 'system/core.js', array('version' => 'auto', 'relative' => true));
+
+		// Add core paths so javascript can have the root path.
+		JFactory::getDocument()->addScriptOptions('system.paths', array('root' => JUri::root(true), 'base', JUri::base(true)));
```
Comment (Quy) [inline libraries/cms/html/behavior.php]: Change to `'base' => JUri::base(true)`?

**Comment by Quy** [comment]
After PR, I still get a js error:
```
Uncaught TypeError: Cannot read property 'system.keepalive' of null
    at Object.e.getOptions (core.js?449d852…:1)
    at HTMLDocument.<anonymous> (keepalive.js?449d852…:1)
```

**Comment by Fedik** [comment]
@andrepereiradasilva I guess there still bug with `Joomla.loadOptions`

there 
```javascript
if (!Joomla.optionsStorage) {
  Joomla.optionsStorage = options;
}
```
need to be:
```javascript
if (!Joomla.optionsStorage) {
  Joomla.optionsStorage = options || {};
}
```
Can you add it here or I do new PR?


@dgt41 I thought that is your PR 😉 
we can update, after this one will be merged

**Comment by Fedik** [comment]
another strange thing, 
with this patch, if cache is enabled, the core.js loaded but there no any `<script type="application/json" class="joomla-script-options">` in the head. 
maybe it somehow related to `JCache::set(get)Workarounds` ?

**Comment by andrepereiradasilva** [comment]
@Fedik
> we can update, after this one will be merged

please do a PR i prefer you do that Joomla.optionsStorage part -  i will test it so it can be RTC and merged fast (op @dgt41 tests it too 😄 )
this one stays in standby until your is merged

> another strange thing,
> with this patch, if cache is enabled, the core.js loaded but there no any <script type="application/json" class="joomla-script-options"> in the head.
> maybe it somehow related to JCache::set(get)Workarounds ?

that is the first problem of the issue that originated this PR see https://github.com/joomla/joomla-cms/issues/16375#issuecomment-305864460
But that's another issue not related to js.

**Comment by Fedik** [comment]
@andrepereiradasilva there is pull #16488 😉 
which should fix:
```
Uncaught TypeError: Cannot read property 'system.keepalive' of null
    at Object.e.getOptions (core.js?449d852…:1)
    at HTMLDocument.<anonymous> (keepalive.js?449d852…:1)
```

**Comment by andrepereiradasilva** [comment]
yeah this work after https://github.com/joomla/joomla-cms/pull/16488 applied

**Comment by Quy** [comment]
I am still getting the js error after both PRs are applied.

**Comment by andrepereiradasilva** [comment]
@Quy did you refresh the js cache?

**Comment by andrepereiradasilva** [comment]
@Quy :  @Fedik solved the issue in the other PR, should be fine now

**Comment by Quy** [comment]
I have tested this item :white_check_mark: successfully on 80d499b0073746cd49225d6300101421e4276181<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/16452">issues.joomla.org/tracker/joomla-cms/16452</a>.</sub>

**Comment by Fedik** [comment]
I have tested this item :white_check_mark: successfully on 80d499b0073746cd49225d6300101421e4276181<br /><br />works as described, with #16488<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/16452">issues.joomla.org/tracker/joomla-cms/16452</a>.</sub>

**Comment by franz-wohlkoenig** [comment]
RTC after two successful tests.

**Pull request merged**